components/brave_news/rust/lib.rs

```diff
@@ -26,10 +27,64 @@ mod ffi {
     }
 
     extern "Rust" {
+        fn strip_html(subject: &str) -> String;
         fn parse_feed_bytes(source: &[u8], output: &mut FeedData) -> bool;
     }
 }
 
+// Note: This function isn't likely to be perfect, but it does guarantee that
+// there will be no HTML tags in the output (as it strips both '<' and '>' from
+// from the output).
+fn strip_html(subject: &str) -> String {
+    let mut output = String::with_capacity(subject.len());
+    let mut depth_tag = 0;
+    let mut depth_comment = 0;
+
+    // We want to know the last 3 characters so we can see if they were --> for
+    // closing a comment.
+    let last_chars_size = 3;
+    let mut last_chars = String::with_capacity(last_chars_size);
+    let mut iterator = subject.chars();
+
+    while let Some(c) = iterator.next() {
+        last_chars.push(c);
+        if last_chars.len() > last_chars_size {
+            last_chars = last_chars.chars().skip(1).collect();
+        }
+
+        if c == '<' {
+            let is_comment = iterator.clone().take(3).collect::<String>() == "!--";
+            if is_comment {
+                depth_comment += 1;
+            } else if depth_comment == 0 {
+                depth_tag += 1;
+            }
+            continue;
+        }
+
+        if c == '>' {
+            if depth_comment > 0 {
+                if last_chars == "-->" {
+                    // If this was the close for a comment, reduce our comment depth.
+                    depth_comment -= 1
```

Comment:
The html spec says [comments can't be nested](https://html.spec.whatwg.org/multipage/syntax.html#comments), so `depth_comment = 0` would be closer to correct parsing here.

Seems like just removing all the `<` and `>` characters would be sufficient to block cross-site scripting attacks? If so this is just about giving a better rendering in plain text, especially in the case of malformed html. The nested tag and comment support removes more than the browser would in that case. e.g. `<![CDATA[<foo>]]>` is stripped entirely instead of rendering as `]]>`. I think erring on the side of caution is fine, though.

Reply:
I think removing `<` on it's own would be sufficient to prevent any XSS attacks, but yeah, I was aiming for a mostly-good-enough plain text conversion.

Do you think it's worth handling `<![CDATA[<foo>]]>` differently? I think stripping it is probably fine for our purposes (though if there's valid content inside we won't print it out).

Reply:
`<![CDATA[...]>` contents are valid xhtml, but html parsers skip them entirely, so it's fine as long as the content is always stripped.